File: WinForms/VB/PivotGridMainDemo/Modules/HitInfo.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraEditors.Controls

Namespace DevExpress.XtraPivotGrid.Demos.Modules

    Public Partial Class HitInfo
        Inherits TutorialControl

        Public Sub New()
            ' This call is required by the Windows Form Designer.
            CreateWaitDialog()
            InitializeComponent()
        End Sub

        Public Overrides ReadOnly Property ViewOptionsControl As PivotGridControl
            Get
                Return pivotGridControl1
            End Get
        End Property

        Public Overrides ReadOnly Property ExportControl As PivotGridControl
            Get
                Return pivotGridControl1
            End Get
        End Property

        Private Sub HitInfo_Load(ByVal sender As Object, ByVal e As EventArgs)
            pivotGridControl1.DataSource = GetNWindData("SalesPerson")
            pivotGridControl1.HeaderImages = HeadersImageList
            pivotGridControl1.ValueImages = CategoryImageList
            fieldSalesPerson.ImageIndex = 0
            fieldCountry.CollapseValue("UK")
            ShowHitInfo(Nothing)
        End Sub

        Private Sub pivotGridControl1_FieldValueImageIndex(ByVal sender As Object, ByVal e As PivotFieldImageIndexEventArgs)
            If e.Field Is fieldCategoryName AndAlso Equals(e.ValueType, PivotGridValueType.Value) Then e.ImageIndex = GetCategoryIndexByName(e.Value)
        End Sub

        '<pivotGridControl1>
        Private Sub pivotGridControl1_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
            ShowHitInfo(pivotGridControl1.CalcHitInfo(New Point(e.X, e.Y)))
        End Sub

        Private Sub pivotGridControl1_MouseLeave(ByVal sender As Object, ByVal e As EventArgs)
            ShowHitInfo(pivotGridControl1.CalcHitInfo(New Point(-1, -1)))
        End Sub

        '<panelControl1>
        Private Sub ShowHitInfo(ByVal info As PivotGridHitInfo)
            ShowCellInfo(info)
            ShowValueInfo(info)
            ShowHeaderInfo(info)
            Dim hintVisible As Boolean = lcgCell.Visibility <> XtraLayout.Utils.LayoutVisibility.Always AndAlso lcgValue.Visibility <> XtraLayout.Utils.LayoutVisibility.Always AndAlso lcgHeadersArea.Visibility <> XtraLayout.Utils.LayoutVisibility.Always
            lciHint.Visibility = If(hintVisible, XtraLayout.Utils.LayoutVisibility.Always, XtraLayout.Utils.LayoutVisibility.Never)
        End Sub

        '</panelControl1>
        '</pivotGridControl1>
        Private Sub ShowCellInfo(ByVal info As PivotGridHitInfo)
            If info Is Nothing OrElse info.CellInfo Is Nothing Then
                lcgCell.Visibility = XtraLayout.Utils.LayoutVisibility.Never
                Return
            End If

            lcgCell.Visibility = XtraLayout.Utils.LayoutVisibility.Always
            textEdit1.Text = info.CellInfo.DisplayText
            textEdit2.Text = String.Format("{0}, {1}", info.CellInfo.ColumnIndex, info.CellInfo.RowIndex)
            textEdit3.Text = String.Format("{0}", info.CellInfo.Value)
            textEdit4.Text = If(info.CellInfo.DataField IsNot Nothing, info.CellInfo.DataField.ToString(), "")
        End Sub

        Private Sub ShowValueInfo(ByVal info As PivotGridHitInfo)
            If info Is Nothing OrElse info.ValueInfo Is Nothing Then
                lcgValue.Visibility = XtraLayout.Utils.LayoutVisibility.Never
                Return
            End If

            lcgValue.Visibility = XtraLayout.Utils.LayoutVisibility.Always
            textEdit5.Text = String.Format("{0}", info.ValueInfo.Value)
            textEdit6.Text = String.Format("{0}", info.ValueInfo.ValueType)
            textEdit7.Text = String.Format("{0}", info.ValueInfo.Field)
        End Sub

        Private Sub ShowHeaderInfo(ByVal info As PivotGridHitInfo)
            If info Is Nothing OrElse info.HeadersAreaInfo Is Nothing Then
                lcgHeadersArea.Visibility = XtraLayout.Utils.LayoutVisibility.Never
                Return
            End If

            lcgHeadersArea.Visibility = XtraLayout.Utils.LayoutVisibility.Always
            textEdit8.Text = String.Format("{0}", info.HeadersAreaInfo.Area)
            textEdit9.Text = String.Format("{0}", info.HeadersAreaInfo.Field)
        End Sub

        Private Sub paddingPanel_Paint(ByVal sender As Object, ByVal e As PaintEventArgs)
        End Sub
    End Class
End Namespace

File: WinForms/CS/PivotGridMainDemo/Modules/HitInfo.cs
using System;
using System.Collections;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraEditors.Controls;
using DevExpress.Data.PivotGrid;

namespace DevExpress.XtraPivotGrid.Demos.Modules {
	public partial class HitInfo : DevExpress.XtraPivotGrid.Demos.TutorialControl {
		public HitInfo() {
			// This call is required by the Windows Form Designer.
			CreateWaitDialog();
			InitializeComponent();
		}

		public override PivotGridControl ViewOptionsControl { get { return pivotGridControl1; }}
		public override PivotGridControl ExportControl { get { return pivotGridControl1; }}

		private void HitInfo_Load(object sender, System.EventArgs e) {
			pivotGridControl1.DataSource = GetNWindData("SalesPerson");
			pivotGridControl1.HeaderImages = HeadersImageList;
			pivotGridControl1.ValueImages = CategoryImageList;
			fieldSalesPerson.ImageIndex = 0;
			fieldCountry.CollapseValue("UK");
			ShowHitInfo(null);
		}


		private void pivotGridControl1_FieldValueImageIndex(object sender, DevExpress.XtraPivotGrid.PivotFieldImageIndexEventArgs e) {
			if(e.Field == fieldCategoryName && Comparer.Equals(e.ValueType, PivotGridValueType.Value))
				e.ImageIndex = GetCategoryIndexByName(e.Value);
		}

        //<pivotGridControl1>
		private void pivotGridControl1_MouseMove(object sender, System.Windows.Forms.MouseEventArgs e) {
			ShowHitInfo(pivotGridControl1.CalcHitInfo(new Point(e.X, e.Y)));
		}

		private void pivotGridControl1_MouseLeave(object sender, System.EventArgs e) {
			ShowHitInfo(pivotGridControl1.CalcHitInfo(new Point(-1, -1)));
		}

        //<panelControl1>
		void ShowHitInfo(PivotGridHitInfo info) {
            ShowCellInfo(info);
            ShowValueInfo(info);
            ShowHeaderInfo(info);
            bool hintVisible = lcgCell.Visibility != XtraLayout.Utils.LayoutVisibility.Always &&
                               lcgValue.Visibility != XtraLayout.Utils.LayoutVisibility.Always &&
                               lcgHeadersArea.Visibility != XtraLayout.Utils.LayoutVisibility.Always;
            lciHint.Visibility = hintVisible ? XtraLayout.Utils.LayoutVisibility.Always : XtraLayout.Utils.LayoutVisibility.Never;
        }
        //</panelControl1>
        //</pivotGridControl1>

        void ShowCellInfo(PivotGridHitInfo info) {
            if(info == null || info.CellInfo == null) {
                lcgCell.Visibility=XtraLayout.Utils.LayoutVisibility.Never;
                return;
            }
            lcgCell.Visibility = XtraLayout.Utils.LayoutVisibility.Always;
            textEdit1.Text = info.CellInfo.DisplayText;
            textEdit2.Text = string.Format("{0}, {1}", info.CellInfo.ColumnIndex, info.CellInfo.RowIndex);
            textEdit3.Text = string.Format("{0}", info.CellInfo.Value);
            textEdit4.Text = info.CellInfo.DataField != null ? info.CellInfo.DataField.ToString() : "";
        }
        void ShowValueInfo(PivotGridHitInfo info) {
            if(info == null || info.ValueInfo== null) {
                lcgValue.Visibility = XtraLayout.Utils.LayoutVisibility.Never;
                return;
            }
            lcgValue.Visibility = XtraLayout.Utils.LayoutVisibility.Always;
            textEdit5.Text = string.Format("{0}", info.ValueInfo.Value);
            textEdit6.Text = string.Format("{0}", info.ValueInfo.ValueType);
            textEdit7.Text = string.Format("{0}", info.ValueInfo.Field);
        }

        void ShowHeaderInfo(PivotGridHitInfo info) {
            if(info == null || info.HeadersAreaInfo == null) {
                lcgHeadersArea.Visibility = XtraLayout.Utils.LayoutVisibility.Never;
                return;
            }
            lcgHeadersArea.Visibility = XtraLayout.Utils.LayoutVisibility.Always;
            textEdit8.Text = string.Format("{0}", info.HeadersAreaInfo.Area);
            textEdit9.Text = string.Format("{0}", info.HeadersAreaInfo.Field);
        }
            private void paddingPanel_Paint(object sender, PaintEventArgs e) {

        }
	}
}

